### PR TITLE
[BYOC] Make CUTLASS BYOC integration 'Collage friendly'  

### PR DESCRIPTION
**Pull request by mbs-octoml**
(See https://discuss.tvm.apache.org/t/byoc-supporting-cutlass-byoc-with-collage/12796/6 for context, which in turn is part of Collage (https://github.com/apache/tvm-rfcs/blob/main/rfcs/0062-collage.md). 

Currently CUTLASS has four entry points:
  - The usual 'partition_for_cutlass' partitioning function, using the standard pattern table and pass machinery (see cutlass/build.py).
  - A 'tune_cutlass_kernels' function which augments CUTLASS partition functions with the results of building and running test kernels (see cutlass/build.py).
  - A 'relay.ext.cutlass' external codegen function which inspects the  turning results and generates a CSourceModule for each partitions (see cutlass/codegen.cc).
  - A 'build_cutlass_kernels_vm' function which runs 'export_library' with all the nvcc compiler options needed to build all the CSourceModules (see cutlass/bild.py).
  
For Collage we'd like CUTLASS to have only two entry points: 'partition_for_cutlass', and 'relay.ext.cutlass' or equivalent. This makes the CUTLASS external codegen integration composable with other integrations, which in turn helps Collage avoid having to understand any external codegen APIs other than the global pattern table and the custom compilation function/pass.  

Collage also tends to end up requiring multiple partitions for the same backend since it is more aggressive at mixing-and-matching smaller sub-graphs between backends. Thus we'd also like to make sure all tuning, generated code and compilation overhead is shared between all such CUTLASS partitions. 

So, in this PR:
- We add all the CUTLASS-specific tuning and compilation options as new Target attributes for the 'external codegen' "cutlass" TargetKind (cutlass/target.cc). The user now has one place to provide those settings, and we've already done the legwork to plumb the target instance.
- We replace 'relay.ext.cutlass' with a 'RelayToTIR' custom pass hook 'CompileForCutlass' (see cutlass/codegen.cc). This pass obviously can see all the CUTLASS partitions in the IRModule, so we can now share tuning results between them all and can be sure to generate a single CSourceModule. The pass can also invoke the compiler to yield a StaticModule, which we've also already done the legwork to support. In this way all CUTLASS-specific steps are handled at once.
- For convenience we supply 'finalize_modules' and 'finalize_modules_vm' which invoke nvcc for final linking (using export_library as usual). However, there's now nothing CUTLASS specific in those helpers other than their overriding of the 'compiler' to be nvcc.
- test_cutlass.py is updated to use the new API.

Though this is a breaking change for existing users of the CUTLASS integration the change is pretty minor, as shown in test_cutlass.py.

### CODE REVIEW

**Comment by mbs-octoml** [comment]
CAUTION: Includes #11619, I'll rebase when possible.

**Comment by mbs-octoml** [comment]
Rebased and ready for review.

**Comment by tmoreau89** [comment]
@apeskov that PR should be of interest to you!

**Comment by masahi** [comment]
> so we can now share tuning results between them all and can be sure to generate a single CSourceModule

If I understand this correctly, this is a huge improvement! Previously, when compiling a model like `bert-large` which has the same workload repeating many times, tuning results are cached but the generated code is not shared. This ends up in a very slow compile and a huge binary blob (> 1G). 

I also like the more polished API. 

cc @Laurawly @comaniac

File: python/tvm/contrib/cutlass/build.py [hunk 1]
```diff
@@ -346,187 +357,271 @@ def tune_cutlass_kernels(
...
+
+    # Recover static library
+    logger.info("Loading compiled CUTLASS code")
+    final_mod = tvm.runtime.load_static_library(lib_path, function_names)
+
+    logger.info("Done with CUTLASS compilation")
+    return final_mod
+
+
+def finalize_modules(lib, lib_path, tmp_dir):
```
Comment (masahi) [inline python/tvm/contrib/cutlass/build.py:575]: Please restore the default arguments for `lib_path` and `tmp_dir`.

File: python/tvm/contrib/cutlass/build.py [hunk 2]
```diff
@@ -346,187 +357,271 @@ def tune_cutlass_kernels(
...
+    lib_path = os.path.join(tmp_dir, lib_path)
+    lib.export_library(lib_path, workspace_dir=tmp_dir, cc="nvcc")
+    return runtime.load_module(lib_path)
 
-    lib_path : string, optional
-        The path to a shared library which will be generated as the result of the build process.
 
-    vmcode_path : string, optional
-        The path where the VM bytecode will be serialized to.
+def finalize_modules_vm(vm_exec, lib_path, tmp_dir):
```
Comment (masahi) [inline python/tvm/contrib/cutlass/build.py:602]: Please restore the default arguments for `lib_path` and `tmp_dir`. Also, `vmcode_path` needs to be provided, to allow loading a pre-compiled model for VM.

**Review by masahi — APPROVED** [review]

**Review by comaniac — APPROVED** [review]
Thanks for the change that makes CUTLASS more practical in TVM :)

File: python/tvm/contrib/cutlass/build.py (hunk 1 above)
Comment (mbs-octoml) [reply python/tvm/contrib/cutlass/build.py:575]: done

File: python/tvm/contrib/cutlass/build.py (hunk 2 above)
Comment (mbs-octoml) [reply python/tvm/contrib/cutlass/build.py:602]: Done. Guess saving the vm image is useful even though a no-op as far as preparing is concerned.

Comment (masahi) [reply python/tvm/contrib/cutlass/build.py:602]: I'm using `vmcode` here in my cutlass byoc benchmark

https://github.com/masahi/tvm-cutlass-eval/blob/master/yolo5/run.py#L133-L136

**Comment by masahi** [comment]
@mbs-octoml Can you remove this line that you just added back? I don't know why I added it before.
https://github.com/apache/tvm/blob/9c16e05020512493d86c249965d08312ab51704c/python/tvm/contrib/cutlass/build.py#L635

**Comment by mbs-octoml** [comment]
removed the unnecessary load

**Pull request merged**